### PR TITLE
feat: Implement Allowlist Module And Native Token Limit Module

### PR DESCRIPTION
**Pull request by Zer0dot**
# Pull Request Checklist

- [x] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [ ] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [x] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [x] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [x] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

Implements features related to the AllowlistModule. TODOS:

- [x] Implement encoding helpers for `onInstall()` and `onUninstall()`
- [x] Test installation and uninstallation
- [x] Test functionality

---

## PR-Codex overview
This PR focuses on introducing new error handling and validation modules in a smart contract system, enhancing the encoding and decoding of data for various modules, and improving test coverage for these functionalities.

### Detailed summary
- Renamed `name` in `EntityIdOverrideError` to `"EntityIdOverrideError"`.
- Added `nativeTokenLimitModule` with encoding methods for install/uninstall data.
- Introduced `allowlistModule` with encoding methods for install/uninstall data.
- Updated tests to include scenarios for `allowlistModule` and `nativeTokenLimitModule`.
- Improved error handling for entity ID validation in `installValidationActions`.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: account-kit/smart-contracts/src/ma-v2/client/client.test.ts [hunk 1]
```diff
@@ -14,8 +14,15 @@ import {
 import { local070Instance } from "~test/instances.js";
 import { setBalance } from "viem/actions";
 import { accounts } from "~test/constants.js";
-import { getDefaultSingleSignerValidationModuleAddress } from "../modules/utils.js";
+import {
+  getDefaultSingleSignerValidationModuleAddress,
+  getDefaultAllowlistModuleAddress,
```
Comment (moldy530) [inline account-kit/smart-contracts/src/ma-v2/client/client.test.ts:19]: ⚠️ **[eslint]** <[@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars)> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'getDefaultAllowlistModuleAddress' is defined but never used.

File: aa-sdk/core/src/errors/client.ts [hunk 2]
```diff
@@ -88,3 +88,19 @@ export class InvalidNonceKeyError extends BaseError {
     );
   }
 }
+
+/**
+ * Error class denoting that the provided entity id is invalid because it's overriding the native entity id.
+ */
+export class EntityIdOverrideError extends BaseError {
+  override name = "InvalidNonceKeyError";
```
Comment (dphilipson) [inline aa-sdk/core/src/errors/client.ts:96]: The class name and string `name` don't match. I'm guessing the string is wrong?

Comment (Zer0dot) [reply aa-sdk/core/src/errors/client.ts]: Addressed in [fb6d3b6](https://github.com/alchemyplatform/aa-sdk/pull/1232/commits/fb6d3b69e2ffc5dac48ffc8f80f0a5d83a3454d3)!

**Review by dphilipson — APPROVED** [review]
Looks great!

File: account-kit/smart-contracts/src/ma-v2/client/client.test.ts [hunk 3]
```diff
@@ -182,6 +192,225 @@ describe("MA v2 Tests", async () => {
...
+            hasPostHooks: false,
+          },
+          initData: hookInstallData,
+        },
+      ],
+    });
+
+    await expect(
+      provider.waitForUserOperationTransaction(installResult)
+    ).resolves.not.toThrowError();
```
[the remark this answers is not in the document]
Comment (Zer0dot) [reply account-kit/smart-contracts/src/ma-v2/client/client.test.ts:244]: Ah makes total sense, let me go ahead and remove these

[the remark this answers is not in the document]
Comment (Zer0dot) [reply account-kit/smart-contracts/src/ma-v2/client/client.test.ts]: [dede43b](https://github.com/alchemyplatform/aa-sdk/pull/1232/commits/dede43bfb19a5bdced0392e840877aed8ffc48e6) done!

**Pull request merged**